test/common/upstream/upstream_impl_test.cc

```diff
@@ -3250,6 +3250,85 @@ TEST_F(ClusterInfoImplTest, Http3BadConfig) {
   EXPECT_THROW_WITH_REGEX(makeCluster(yaml), EnvoyException,
                           "HTTP3 requires a QuicUpstreamTransport transport socket: name.*");
 }
+
+TEST_F(ClusterInfoImplTest, Http3Alpn) {
+  const std::string yaml = TestEnvironment::substitute(R"EOF(
+    name: name
+    connect_timeout: 0.25s
+    type: STRICT_DNS
+    lb_policy: MAGLEV
+    load_assignment:
+        endpoints:
+          - lb_endpoints:
+            - endpoint:
+                address:
+                  socket_address:
+                    address: foo.bar.com
+                    port_value: 443
+    transport_socket:
+      name: envoy.transport_sockets.tls
+      typed_config:
+        "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
+        common_tls_context:
+          tls_certificates:
+          - certificate_chain:
+              filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/san_uri_cert.pem"
+            private_key:
+              filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/san_uri_key.pem"
+          validation_context:
+            trusted_ca:
+              filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/ca_cert.pem"
+            match_subject_alt_names:
+            - exact: localhost
+            - exact: 127.0.0.1
+    transport_socket_matches:
+      - name: use_quic
+        match:
+          use_quic: true
+        transport_socket:
+          name: envoy.transport_sockets.quic
+          typed_config:
+            "@type": type.googleapis.com/envoy.extensions.transport_sockets.quic.v3.QuicUpstreamTransport
+            upstream_tls_context:
+              common_tls_context:
+                tls_certificates:
+                - certificate_chain:
+                    filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/san_uri_cert.pem"
+                  private_key:
+                    filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/san_uri_key.pem"
+                validation_context:
+                  trusted_ca:
+                    filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/ca_cert.pem"
+                  match_subject_alt_names:
+                  - exact: localhost
+                  - exact: 127.0.0.1
```

Comment:
In looking at this in config, I agree that it seems tremendously awful that common_tls_context is going to get repeated in both cases.

What are your thoughts on making the quic transport socket be a wrapped transport socket that is aware of how to do TCP fallback to a standard TLS/TCP transport socket? If we made the combined configuration explicit it would be a lot less verbose, we could do away with the matching entirely, and just have the fallback process be a bit more built-in?

Like I could imagine the existing quic transport socket having a tcp_fallback with an explicit type of `envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext`. I wonder if it would even need any configuration since it could just be documented to share the same common TLS context until someone actually asks for there to be different config?

Then I wonder if instead we would do a few things to the transport socket API itself:
1) Do a bit more with "feature capabilities" so calling code can check for correct configuration (we already started to do this with the starttls functionality iirc).
2) Make the createTransportSocket() function take optional feature capability bits that would allow calling code to control what is returned.

I think if we do (1) and (2) along with the wrapped transport this would be way cleaner. WDYT? cc @lizan and @lambdai who might also have some ideas.

Reply:
Agreed it is ugly the common_tls_context is get repeated. +1 to make (1) and (2) happen.

Reply:
Yeah, I 100% like having QUIC have auto-failover - we'll need it all the time for internet access (both for mobile and DFP) and can either configure it off or lazily create for folks on intranets where QUIC won't fail.

So just digging on 2, your preference would be to have a single factory which can create H3 or H2 sockets, and we basically make tcp vs udp one of the transport options?  I'll prototype that and ping back once it's ready for a look.

This won't change anything for quic downstream config but ccl @danzh2010 just so it's on her radar.

Reply:
Looking a bit more I think if we go with that, we'll have to either create transport socket options or wrap transport socket options for all calls, which feels a bit unwieldy.  I think barring objections I'd like to have UDP be a special case, so we createTransportSocket(transport_socket_options, protocol = TCP).
I'll give that a shot and see how it turns out.